co-log/src/Colog/Rotation.hs

```diff
@@ -0,0 +1,112 @@
+    => Limit
+    -- ^ Max allowed file size in bytes
+    -> Limit
+    -- ^ Max allowed number of files to have
+    -> FilePath
+    -- ^ File path to log
+    -> (FilePath -> IO ())
+    -- ^ What to do with old files; pass @removeFile@ here for deletion
+    -> (Handle -> LogAction m msg)
+    -- ^ Action that writes to file handle
+    -> (LogAction m msg -> IO r)
+    -- ^ Continuation action
+    -> IO r
+withLogRotation sizeLimit filesLimit path cleanup mkAction cont = do
+    -- TODO: figure out how to use bracket to safely manage
+    -- possible exceptions
+    handle <- openFile path AppendMode
+    handleRef <- newIORef handle
+    cont $ rotationAction handleRef
+  where
+    rotationAction :: IORef Handle -> LogAction m msg
+    rotationAction refHandle
+      = LogAction $ \msg -> do
+        handle <- liftIO $ readIORef refHandle
+        unLogAction (mkAction handle) msg
+
+        whenM
+          (liftIO $ isFileSizeLimitReached sizeLimit handle)
+          (cleanupAndRotate refHandle)
+    cleanupAndRotate :: IORef Handle -> m ()
+    cleanupAndRotate refHandle = liftIO $ do
+      h <- readIORef refHandle
+      hClose h
+      maxN <- maxFileIndex path
+      renameFileToNumber (succ maxN) path
+      oldFiles <- getOldFiles filesLimit path
+      mapM_ cleanup oldFiles
+      newHandle <- openFile path AppendMode
+      modifyIORef' refHandle (const newHandle)
+
+isLimitedBy :: Integer -> Limit -> Bool
+isLimitedBy _ Unlimited = False
+isLimitedBy size (LimitTo limit) | size <= 0 = False
+                                 | otherwise = limit > (fromInteger size :: Natural)
+
+isFileSizeLimitReached :: Limit -> Handle -> IO Bool
+isFileSizeLimitReached limit handle = do
+  fileSize <- hFileSize handle
+  pure $ isLimitedBy fileSize limit
+
+-- if you have files node.log.0, node.log.1 and node.log.2 then this function
+-- will return `2` if you give it `node.log`
+maxFileIndex :: FilePath -> IO Int
+maxFileIndex path = do
+  files <- D.listDirectory (POS.takeDirectory path)
+  let logFiles = filter (== POS.takeBaseName path) files
+  let maxFile = getMax . foldMap Max <$> nonEmpty (mapMaybe logFileIndex logFiles)
+  pure $ fromMaybe 0 maxFile
+
+-- given number 4 and path `node.log` renames file `node.log` to `node.log.4`
+renameFileToNumber :: Int -> FilePath -> IO ()
+renameFileToNumber n path = D.renameFile path (path POS.<.> show n)
+
+-- if you give it name like `node.log.4` then it returns `Just 4`
+logFileIndex :: FilePath -> Maybe Int
+logFileIndex path = nonEmpty (POS.takeExtension path) >>= readMaybe . tail
+
+-- creates list of files with indices who are older on given Limit than the latest one
+getOldFiles :: Limit -> FilePath -> IO [FilePath]
+getOldFiles limit path = do
+    currentMaxN <- maxFileIndex path
+    files <- D.listDirectory (POS.takeDirectory path)
+    let tuple = map (\a -> (a, toInteger <$> logFileIndex a)) files
+    pure $ map fst $ filter (maybe False (isOldFile currentMaxN) . snd) tuple
```

Comment:
I feel like this two lines of code can be simplified. For example, you can use `mapMaybe` function to get rid of `maybe False` and `filter`

Reply:
I would need a function like:
`sequenceC :: (Traversable t, Monad m) => t (a, m b) -> t (m (a, b))`
I could write it myself, but it would complicate the code just to clean these two lines. Any idea?

Reply:
Okay, it's usually helps to write types of functions to see how it can be simplified :slightly_smiling_face: Usually in such situations Haskell reminds this child game where you have objects of different shapes and you need to put them inside box where holes of those shapes should match :slightly_smiling_face:  

Context:

```haskell
files :: [FilePath]
logFileIndex :: FilePath -> Maybe Natural
isOldFile maxN :: Natural -> Bool
mapMaybe :: (a -> Maybe b) -> [a] -> [b]

<goal> :: [FilePath] -- by `boolean predicate
```

The idea is to implement couple functions:

```haskell
takeFileIndex  :: FilePath -> Maybe (FilePath, Natural)
guardFileIndex :: (FilePath, Natural) -> Maybe FilePath
-- should be quite easy using the existing functions
```

Then you will have:

```haskell
mapMaybe (takeFileIndex >=> guardFileIndex) files :: [FilePath]
```

No need to play with `Bool`, I personally like `mapMaybe` function, it's really powerful.

Reply:
Wow this is amazing. I'm seeing a pattern in these refactorings: when things look like they can't compose, break them down :)